### PR TITLE
Homework - Cars

### PR DESCRIPTION
**Pull request by anraf1001**
Fixes:
- checking if gear that user wants to set is OK
- `accelerate()` only changes speed when parameter >= 0
- added `speed_` variable to Car class that stores actual car speed
- `changeEngine()` method checks if proper engine type is given as parameter
- `changeEngine()` method checks if car's speed equals 0
- engines in cars are protected
- power, batteryCapacity, capacity, gears, currentGear are protected/private

### CODE REVIEW

File: ElectricCar.hpp [hunk 1]
```diff
@@ -1,16 +1,22 @@
...
-    ~ElectricCar();
-    void turnLeft();
-    void turnRight();
-    void brake();
-    void accelerate(int speed);
+
+    // Override from Car
+    ~ElectricCar() override;
+    void restore() override;
+    void changeEngine(Engine* engine) override;
```
Comment (WBigaj) [inline ElectricCar.hpp:13]: Isn't better to take specific type of engine for car. Like electric engine for electric car?

File: ElectricCar.cpp [hunk 2]
```diff
@@ -1,16 +1,45 @@
...
+int ElectricCar::getElectricPower() const {
+    return electricEngine_->getPower();
+}
+
+void ElectricCar::changeEngine(Engine* engine) {
+    if (speed_ == 0) {
+        if (typeid(*engine) != typeid(ElectricEngine)) {
+            std::cout << "Wrong type of engine\n";
+            return;
+        }
```
Comment (WBigaj) [inline ElectricCar.cpp:37]: Would type check be necessary if electric car would take electric engine type?

File: PetrolCar.cpp [hunk 3]
```diff
@@ -1,16 +1,50 @@
...
+void PetrolCar::refuel() {
+    std::cout << __FUNCTION__ << std::endl;
+}
+
+void PetrolCar::changeEngine(Engine* engine) {
+    if (speed_ == 0) {
+        if (typeid(*engine) != typeid(PetrolEngine)) {
+            std::cout << "Wrong type of engine\n";
+            return;
+        }
```
Comment (WBigaj) [inline PetrolCar.cpp:34]: Similar here isn't better to take petrol engine for petrol car?

**Review by WBigaj — COMMENTED** [review]
I have checked your code and have some questions.

File: ElectricCar.hpp (hunk 1 above)
Comment (MaciejSalamonski) [reply ElectricCar.hpp:13]: What if it's a hybrid car?

Comment (MaciejSalamonski) [reply ElectricCar.hpp:13]: > Isn't better to take specific type of engine for car. Like electric engine for electric car?
What if it's a hybrid car?

**Review by MaciejSalamonski — APPROVED** [review]
Nice code! Did you check for memory leaks? I think they might be somewhere. Have a good day!

**Comment by anraf1001** [comment]
> Nice code! Did you check for memory leaks? I think they might be somewhere. Have a good day!

You were right. There was some memory leaks but now everything is fixed. Thanks for review

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
4 XP za wyjątek InvalidGear w Cars (3521)
7 * 2 XP za poprawkę interfejsu w Cars (3522)
10 XP za testy w Cars (3523)
3 * 2 XP za punktualność (1119)

{ "requirement_ids": [3521, 3523, 3522, 3522, 3522, 3522, 3522, 3522, 3522, 1119, 1119, 1119] }

**Pull request closed without merging**